sdk/core/azure-core/inc/azure/core/md5.hpp

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Utility functions to help compute the MD5 hash value for the input binary data.
+ */
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+namespace Azure { namespace Core {
+
+  /**
+   * @brief Represents the class for the MD5 hash function which maps binary data of an arbitrary
+   * length to small binary data of a fixed length.
+   */
+  class Md5 {
+  public:
+    /**
+     * @brief Construct a default instance of @Md5.
+     */
+    Md5();
+
+    /**
+     * @brief Cleanup any state when destroying the instance of @Md5.
+     */
+    ~Md5();
+
+    /**
+     * @brief Used to append parital binary input data to compute the hash in a streaming fashion.
+     * @param data The input vector that contains the current block of binary data that is used for
+     * hash calculation.
+     * @remark Once all the data has been added, call #Digest() to get the computed hash value.
+     */
+    void Update(const std::vector<uint8_t>& data);
+
+    /**
+     * @brief Computes the hash value of all the binary input data appended to the instance so far.
+     * @return The computed MD5 hash value corresponding to the input provided.
+     * @remark Use #Update() to add more partial data before calling this function.
+     */
+    std::vector<uint8_t> Digest() const;
```

Comment:
@JeffreyRichter - are we happy with these names? It took me a bit of time to realize what `Update()` and `Digest()` are supposed to mean since they are not very sefl-documenting. `Hash()` is the single-shot static API, whereas `Update()` is meant to append partial data and once it's all done, `Digest()` is called to compute the hash of everything that was appended so far.

**Should these be something like:**
`AppendData()`
`ComputeHash()` or `HashFinal()`

Also, I only kept a single overload of `Hash()` which accepts a `vector<uint8_t>`. The previous implementation had one with `ptr, length` and another with `std::string.

I don't like exposing pointer-based overloads, especially since the customer can use the `ptr, length` and create a vector out of it quite easily anyway. Let me know if you disagree.

From what I understood, the primary use case of the `Hash()` function was so users would compute the hash on blob data, which is generally binary/bytes. Do we still want to expose an overload that computes the MD5 hash for text string (i.e. is there a known scenario for it already), or can we defer adding that until there is a customer request?

Reply:
storage sdk is supposed to handle large data, way larger than memory. If you only keep `Hash()`, this MD5 thing is almost useless to storage customers.

Reply:
Also, we cannot make any assumption about how our customers are reading the large data. They may use `std::fsteram` or `mmap` or `FILE*`, if you do not support `char*` or `uint8_t*`, customers will have to make a copy of the data to call MD5 function. Data copy speed on a typical VM is like 1~2GB maybe up to several GB per second. this might be unacceptable to some customers.

Reply:
Good point, I didn't realize data copies were happening when creating a vector from `ptr, length`. I will add that overload.

Reply:
What if Core define an abstract Hash base class that MD5 (and other potential hash algorithms derived from).
struct Hash {
   void Append(const std::vector<uint8_t>& data) = 0
   std::vector<uint8_t> Sum(const std::vector<uint8_t>& data) = 0
   int GetSize() = 0            // The Md5 derived class returns 16 
   int GetBlockSize() = 0   // The Md5 derived class returns 64

/// We could add a reset method to reset the Hash object but I'd postpone this to a future version should it ever be needed
}

I don't think we need the static convenience helper method; if we do, we could always add it later. But, it's not hard to "new Md5Hash().Sum(...)"

Reply:
> What if Core define an abstract Hash base class that MD5

I think you bring up some valuable ideas that I would like to pursue to make sure our public surface area is solid. However, a redesign of the hashing APIs is out of the scope of this PR, so I am looking to first move the capability into core, and I'll follow up with you offline and discuss the final shape we want separately.

Reply:
Resolved in https://github.com/Azure/azure-sdk-for-cpp/pull/1632